lightning/src/ln/chan_utils.rs

```diff
@@ -550,11 +556,441 @@ pub fn build_htlc_transaction(prev_hash: &Txid, feerate_per_kw: u32, contest_del
+pub struct HolderCommitmentTransaction {
+	/// The non-party-specific transaction information
+	pub info: CommitmentTransaction,
+	/// Our counterparty's signature for the transaction
+	pub counterparty_sig: Signature,
+	/// All non-dust counterparty HTLC signatures, in the order they appear in the transaction
+	pub counterparty_htlc_sigs: Vec<Signature>,
+}
+
+impl PartialEq for HolderCommitmentTransaction {
+	// We dont care whether we are signed in equality comparison
+	fn eq(&self, o: &Self) -> bool {
+		self.info == o.info
+	}
+}
+
+impl_writeable!(HolderCommitmentTransaction, 0, {
+	info, counterparty_sig, counterparty_htlc_sigs
+});
+
+impl HolderCommitmentTransaction {
+	#[cfg(test)]
+	pub fn dummy() -> Self {
+		let dummy_key = PublicKey::from_secret_key(&Secp256k1::new(), &SecretKey::from_slice(&[42; 32]).unwrap());
+		let dummy_sig = Secp256k1::new().sign(&secp256k1::Message::from_slice(&[42; 32]).unwrap(), &SecretKey::from_slice(&[42; 32]).unwrap());
+
+		HolderCommitmentTransaction {
+			info: CommitmentTransaction {
+				commitment_number: 0,
+				to_broadcaster_value_sat: 0,
+				to_countersignatory_value_sat: 0,
+				feerate_per_kw: 0,
+				htlcs: Vec::new(),
+				keys: TxCreationKeys {
+					per_commitment_point: dummy_key.clone(),
+					revocation_key: dummy_key.clone(),
+					broadcaster_htlc_key: dummy_key.clone(),
+					countersignatory_htlc_key: dummy_key.clone(),
+					broadcaster_delayed_payment_key: dummy_key.clone(),
+				}
+			},
+			counterparty_sig: dummy_sig,
+			counterparty_htlc_sigs: Vec::new()
+		}
+	}
+	pub(crate) fn txid<T: secp256k1::Signing + secp256k1::Verification>(&self, channel_parameters: &DirectedChannelTransactionParameters, secp_ctx: &Secp256k1<T>) -> Txid {
+		self.info.txid(channel_parameters, secp_ctx)
+	}
+}
+
+impl HolderCommitmentTransaction {
+	pub(crate) fn to_holder_commitment_tx<T: secp256k1::Signing + secp256k1::Verification>(&self, channel_parameters: &ChannelTransactionParameters, secp_ctx: &Secp256k1<T>) -> HolderCommitmentTransactionForSigning {
+		let opt_sigs: Vec<Option<Signature>> = self.counterparty_htlc_sigs.iter().map(|s| Some(s.clone())).collect();
+		let mut htlcs = self.info.htlcs.clone();
+		let htlcs_with_sig = htlcs.drain(..).zip(opt_sigs).collect();
+		self.info.to_holder_commitment_tx(self.counterparty_sig, htlcs_with_sig, &channel_parameters.to_directed(true), secp_ctx)
+	}
+}
+
+/// This class tracks the information needed to build a commitment transaction and to actually
+/// build it and sign.  It is used for holder transactions that we sign only when needed
+/// and for transactions we sign for the counterparty.
+///
+/// This class can be used inside a signer implementation to generate a signature given the relevant
+/// secret key.
+#[derive(PartialEq)]
+#[derive(Clone)]
+pub struct CommitmentTransaction {
+	/// The backwards-counting commitment number
+	pub commitment_number: u64,
+	/// The value to be sent to the broadcaster
+	pub to_broadcaster_value_sat: u64,
```

Comment:
Shouldn't the explicit values be replaced with the set of dust HTLCs so that an external signer can at least see the dust HTLCs, even if they aren't included?

Reply:
the only interesting policy on dust HTLCs that I came up with is a limit on their aggregate value, which can be enforced without knowing their details.  any other policies you can think of?

Reply:
I still don't see how the signer would use this information.  Here are my thoughts:

- the info in the dust `HTLCOutputInCommitment` cannot be verified by the signer.  It could have been invented by a compromised node.
- as opposed to the non-dust HTLCs, the info is not going on-chain and is not signed by the counterparty, so doesn't describe any recourse we have

So it seems like the only thing the signer can do with this is sum the individual values and see everything adds up, but that doesn't seem to buy us any additional security, since we already know the result.

The one policy I can see related to this is a limit on the sum of dust HTLCs, which can be computed from the existing fields already.

Am I missing something?

Reply:
I was thinking that a signer may use information about dust HTLCs in one channel to correlate with non-dust HTLCs in another channel (with a different dust limit) that represent forwarding of the same payment. Or possibly correlate HTLCs in the inbound/outbound commitment transactions in the same channel, since IIRC both could have different dust limits.

I admit I don't in any way have  a coherent threat model for how using this data would be helpful, only that it seems strange to not pass it to the signer.